packages/langium/src/utils/promise-util.ts

```diff
@@ -73,7 +73,119 @@ export async function interruptAndCheck(token: CancellationToken): Promise<void>
         lastTick = current;
         await delayNextTick();
     }
-    if (token. isCancellationRequested) {
+    if (token.isCancellationRequested) {
         throw OperationCancelled;
     }
 }
+
+/**
+ * Utility class to allow for mutually exclusive read and write access.
+ *
+ * Can perform multiple read accesses at the same time, but only one write access.
+ */
+export class ReadWriteMutex {
+
+    private readCount = 0;
+    private writeCount = 0;
+
+    private writeLock: Deferred = new Deferred().resolve();
+    private readLock: Deferred = new Deferred().resolve();
+
+    private previousWriteAction?: Promise<void>;
+
+    /**
+     * Performs a single async write action, like initializing the workspace or processing document changes.
+     * Only one action will be executed at a time.
+     *
+     * If a write action is queued up while a read action is being performed, the write action will wait until all read actions have completed.
+     */
+    async write(action: () => Promise<void>): Promise<void> {
+        // Lock reading first, so that consequent read requests don't have priority over this write request
+        this.lockRead();
+        // Then await that every read previous action has completed
+        await this.readLock.promise;
+        // Append the new action to the previous action. We usually don't have to wait for long, as the previous write action
+        // 1. has either completed
+        // 2. has been cancelled due to the new write request
+        this.previousWriteAction = (this.previousWriteAction ?? Promise.resolve()).then(() =>
+            action().catch(err => {
+                if (!isOperationCancelled(err)) {
+                    console.error('Error: ', err);
+                }
+            }).finally(() => {
+                this.unlockRead();
+            })
+        );
+    }
+
+    /**
+     * Call this to lock this mutex before a write operation.
+     */
+    private lockRead(): void {
+        if (this.writeCount === 0) {
+            this.writeLock.resolve();
+            this.writeLock = new Deferred();
+        }
+        this.writeCount++;
+    }
+
+    /**
+     * Call this to unlock this mutex after a write operation.
+     */
+    private unlockRead(): void {
+        this.writeCount--;
```

Comment:
I believe there's a potential issue here if `unlockRead` is called when no previous writeLock has been acquired. That would drop the writeCount below zero, and result in some misbehavior with subsequent locks. It would be a good idea to do something to guard against this, either throwing something or silently preventing such cases. This depends on the implementation of course, and a good usage of it may never trip the issue, but it would still be prudent to take care of.